jumanji/environments/routing/sokoban/env.py

```diff
@@ -0,0 +1,587 @@
+            jnp.all(action == -1),
+            lambda: (state.variable_grid, state.agent_location),
+            lambda: self.move_agent(state.variable_grid, action, state.agent_location),
+        )
+
+        next_state = State(
+            key=state.key,
+            fixed_grid=state.fixed_grid,
+            variable_grid=next_variable_grid,
+            agent_location=next_agent_location,
+            step_count=state.step_count + 1,
+        )
+
+        target_reached = self.level_complete(next_state)
+        time_limit_exceeded = next_state.step_count >= self.time_limit
+
+        #done = target_reached | time_limit_exceeded
+        done = jnp.logical_or(target_reached, time_limit_exceeded)
+
+        reward = jnp.asarray(self.reward_fn(state, action, next_state), float)
+
+        observation = self._state_to_observation(next_state)
+
+        extras = self._get_extras(next_state)
+
+        timestep = jax.lax.cond(
+            done,
+            lambda: termination(
+                reward=reward,
+                observation=observation,
+                extras=extras,
+            ),
+            lambda: transition(
+                reward=reward,
+                observation=observation,
+                extras=extras,
+            ),
+        )
+
+        return next_state, timestep
+
+    def observation_spec(self) -> specs.Spec[Observation]:
+        """
+        Returns the specifications of the observation of the `Sokoban`
+        environment.
+
+        Returns:
+            specs.Spec[Observation]: The specifications of the observations.
+        """
+        grid = specs.BoundedArray(
+            shape=(self.num_rows, self.num_cols, 2),
+            dtype=jnp.uint8,
+            minimum=0,
+            maximum=4,
+            name="grid",
+        )
+        step_count = specs.Array((), jnp.int32, "step_count")
+        return specs.Spec(
+            Observation,
+            "ObservationSpec",
+            grid=grid,
+            step_count=step_count,
+        )
+
+    def action_spec(self) -> specs.DiscreteArray:
+        """
+        Returns the action specification for the Sokoban environment.
+        There are 4 actions: [0,1,2,3] -> [Up, Down, Left, Right].
+
+        Returns:
+            specs.DiscreteArray: Discrete action specifications.
+        """
+        return specs.DiscreteArray(4, name="action", dtype=jnp.int32)
+
+    def _state_to_observation(self, state: State) -> Observation:
+        """Maps an environment state to an observation.
+
+        Args:
+            state: `State` object containing the dynamics of the environment.
+
+        Returns:
+            The observation derived from the state.
+        """
+
+        total_grid = jnp.concatenate(
+            (
+                jnp.expand_dims(state.variable_grid, axis=-1),
+                jnp.expand_dims(state.fixed_grid, axis=-1),
+            ),
+            axis=-1,
+        )
```

Comment:
NIT: I think there is a one-line command to do this, e.g. `jnp.stack([state.variable_grid, state.fixed_grid], dim=-1)` or something.